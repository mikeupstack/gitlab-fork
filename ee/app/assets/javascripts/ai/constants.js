import { s__, sprintf } from '~/locale';

export const i18n = {
  GENIE_TOOLTIP: s__('AI|What does the selected code mean?'),
  GENIE_NO_CONTAINER_ERROR: s__("AI|The container element wasn't found, stopping AI Genie."),
  GENIE_CHAT_TITLE: s__('AI|Code Explanation'),
  GENIE_CHAT_CLOSE_LABEL: s__('AI|Close the Code Explanation'),
  GENIE_CHAT_LEGAL_NOTICE: sprintf(
    s__(
      'AI|You are not allowed to copy any part of this output into issues, comments, GitLab source code, commit messages, merge requests or any other user interface in the %{gitlabOrg} or %{gitlabCom} groups.',
    ),
    { gitlabOrg: '<code>/gitlab-org</code>', gitlabCom: '<code>/gitlab-com</code>' },
    false,
  ),
  GENIE_CHAT_LEGAL_GENERATED_BY_AI: s__('AI|Responses generated by AI'),
  REQUEST_ERROR: s__('AI|Something went wrong. Please try again later'),
  EXPERIMENT_BADGE: s__('AI|Experiment'),
  EXPLAIN_CODE_PROMPT: s__(
    'AI|Explain the code from %{filePath} in human understandable language presented in Markdown format. In the response add neither original code snippet nor any title. `%{text}`',
  ),
  TOO_LONG_ERROR_MESSAGE: s__(
    'AI|There is too much text in the chat. Please try again with a shorter text.',
  ),
  GENIE_CHAT_PROMPT_PLACEHOLDER: s__('AI|You can ask AI for more information.'),
};
export const TOO_LONG_ERROR_TYPE = 'too-long';
export const AI_GENIE_DEBOUNCE = 300;
export const GENIE_CHAT_MODEL_ROLES = {
  user: 'user',
  system: 'system',
  assistant: 'assistant',
};

export const FEEDBACK_OPTIONS = [
  {
    title: s__('AI|Helpful'),
    icon: 'thumb-up',
    value: 'helpful',
  },
  {
    title: s__('AI|Unhelpful'),
    icon: 'thumb-down',
    value: 'unhelpful',
  },
  {
    title: s__('AI|Wrong'),
    icon: 'status_warning',
    value: 'wrong',
  },
];

const MODEL_MAX_TOKENS = 4096; // max for 'gpt-3.5-turbo' model as per https://platform.openai.com/docs/models/gpt-3-5
export const MAX_RESPONSE_TOKENS = 300;
export const TOKENS_THRESHOLD = MODEL_MAX_TOKENS * 0.85; // We account for 15% fault in tokens calculation

export const EXPLAIN_CODE_TRACKING_EVENT_NAME = 'explain_code_blob_viewer';
