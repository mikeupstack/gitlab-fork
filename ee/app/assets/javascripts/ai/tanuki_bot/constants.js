import { s__ } from '~/locale';

export const MESSAGE_TYPES = {
  USER: 'user',
  TANUKI: 'tanuki',
};

export const SOURCE_TYPES = {
  HANDBOOK: {
    value: 'handbook',
    icon: 'book',
  },
  DOC: {
    value: 'doc',
    icon: 'documents',
  },
  BLOG: {
    value: 'blog',
    icon: 'list-bulleted',
  },
};

export const ERROR_MESSAGE = s__(
  'TanukiBot|There was an error communicating with GitLab Chat. Please try again later.',
);

export const TANUKI_BOT_FEEDBACK_ISSUE_URL = 'https://gitlab.com/gitlab-org/gitlab/-/issues/408527';
