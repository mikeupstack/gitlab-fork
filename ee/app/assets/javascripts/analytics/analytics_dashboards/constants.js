import { s__, __ } from '~/locale';

export const I18N_DASHBOARD_LIST_TITLE = s__('Analytics|Analytics dashboards');
export const I18N_DASHBOARD_LIST_DESCRIPTION = s__(
  'Analytics|Dashboards are created by editing the projects dashboard files.',
);
export const I18N_DASHBOARD_LIST_LEARN_MORE = __('Learn more.');
export const I18N_DASHBOARD_LIST_VISUALIZATION_DESIGNER = s__('Analytics|Visualization Designer');
export const I18N_DASHBOARD_LIST_VISUALIZATION_DESIGNER_CUBEJS_ERROR = s__(
  'Analytics|An error occurred while loading data',
);

export const I18N_ALERT_NO_POINTER_TITLE = s__('Analytics|Custom dashboards');
export const I18N_ALERT_NO_POINTER_BUTTON = s__('Analytics|Configure Dashboard Project');
export const I18N_ALERT_NO_POINTER_DESCRIPTION = s__(
  'Analytics|To create your own dashboards, first configure a project to store your dashboards.',
);

export const I18N_DASHBOARD_NOT_FOUND_TITLE = s__('Analytics|Dashboard not found');
export const I18N_DASHBOARD_NOT_FOUND_DESCRIPTION = s__(
  'Analytics|No dashboard matches the specified URL path.',
);
export const I18N_DASHBOARD_NOT_FOUND_ACTION = s__('Analytics|View available dashboards');

export const VISUALIZATION_TYPE_FILE = 'yml';
export const VISUALIZATION_TYPE_BUILT_IN = 'builtin';

export const EVENTS_TYPES = ['pageViews', 'featureUsages', 'clickEvents', 'events'];

export function isTrackedEvent(eventType) {
  return EVENTS_TYPES.includes(eventType);
}

export const PANEL_VISUALIZATION_HEIGHT = '600px';

export const PANEL_DISPLAY_TYPES = {
  DATA: 'data',
  VISUALIZATION: 'visualization',
  CODE: 'code',
};

export const PANEL_DISPLAY_TYPE_ITEMS = [
  {
    type: PANEL_DISPLAY_TYPES.DATA,
    icon: 'table',
    title: s__('Analytics|Data'),
  },
  {
    type: PANEL_DISPLAY_TYPES.VISUALIZATION,
    icon: 'chart',
    title: s__('Analytics|Visualization'),
  },
  {
    type: PANEL_DISPLAY_TYPES.CODE,
    icon: 'code',
    title: s__('Analytics|Code'),
  },
];

export const MEASURE_COLOR = '#00b140';
export const DIMENSION_COLOR = '#c3e6cd';

export const EVENTS_DB_TABLE_NAME = 'TrackedEvents';
export const SESSIONS_TABLE_NAME = 'Sessions';

export const ANALYTICS_FIELD_CATEGORIES = [
  {
    name: s__('Analytics|Pages'),
    category: 'pages',
  },
  {
    name: s__('Analytics|Users'),
    category: 'users',
  },
];

export const ANALYTICS_FIELDS = [
  {
    name: s__('Analytics|URL'),
    category: 'pages',
    dbField: 'url',
    icon: 'documents',
  },
  {
    name: s__('Analytics|Page Path'),
    category: 'pages',
    dbField: 'docPath',
    icon: 'documents',
  },
  {
    name: s__('Analytics|Page Title'),
    category: 'pages',
    dbField: 'pageTitle',
    icon: 'documents',
  },
  {
    name: s__('Analytics|Page Language'),
    category: 'pages',
    dbField: 'docEncoding',
    icon: 'documents',
  },
  {
    name: s__('Analytics|Host'),
    category: 'pages',
    dbField: 'docHost',
    icon: 'documents',
  },
  {
    name: s__('Analytics|Referer'),
    category: 'users',
    dbField: 'referer',
    icon: 'user',
  },
  {
    name: s__('Analytics|Language'),
    category: 'users',
    dbField: 'userLanguage',
    icon: 'user',
  },
  {
    name: s__('Analytics|Viewport'),
    category: 'users',
    dbField: 'vpSize',
    icon: 'user',
  },
  {
    name: s__('Analytics|Browser Family'),
    category: 'users',
    dbField: 'parsedUaUaFamily',
    icon: 'user',
  },
  {
    name: s__('Analytics|Browser'),
    category: 'users',
    dbField: ['parsedUaUaFamily', 'parsedUaUaVersion'],
    icon: 'user',
  },
  {
    name: s__('Analytics|OS'),
    category: 'users',
    dbField: 'parsedUaOsFamily',
    icon: 'user',
  },
  {
    name: s__('Analytics|OS Version'),
    category: 'users',
    dbField: ['parsedUaOsFamily', 'parsedUaOsVersion'],
    icon: 'user',
  },
];
