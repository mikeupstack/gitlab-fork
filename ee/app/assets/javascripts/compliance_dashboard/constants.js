import { __, s__ } from '~/locale';

export const INPUT_DEBOUNCE = 500;

export const CUSTODY_REPORT_PARAMETER = 'commit_sha';

export const DRAWER_AVATAR_SIZE = 24;

export const DRAWER_MAXIMUM_AVATARS = 20;

export const GRAPHQL_PAGE_SIZE = 20;

export const COMPLIANCE_DRAWER_CONTAINER_CLASS = '.content-wrapper';

const APPROVED_BY_COMMITTER = 'APPROVED_BY_COMMITTER';
const APPROVED_BY_INSUFFICIENT_USERS = 'APPROVED_BY_INSUFFICIENT_USERS';
const APPROVED_BY_MERGE_REQUEST_AUTHOR = 'APPROVED_BY_MERGE_REQUEST_AUTHOR';

export const MERGE_REQUEST_VIOLATION_MESSAGES = {
  [APPROVED_BY_COMMITTER]: s__('ComplianceReport|Approved by committer'),
  [APPROVED_BY_INSUFFICIENT_USERS]: s__('ComplianceReport|Less than 2 approvers'),
  [APPROVED_BY_MERGE_REQUEST_AUTHOR]: s__('ComplianceReport|Approved by author'),
};

export const DEFAULT_SORT = 'SEVERITY_LEVEL_DESC';

export const DEFAULT_PAGINATION_CURSORS = {
  before: null,
  after: null,
  first: GRAPHQL_PAGE_SIZE,
};

export const BRANCH_FILTER_OPTIONS = {
  allBranches: __('All branches'),
  allProtectedBranches: __('All protected branches'),
};

export const FRAMEWORK_BADGE_SIZE_SM = 'sm';
export const FRAMEWORK_BADGE_SIZE_MD = 'md';
export const FRAMEWORK_BADGE_SIZES = [FRAMEWORK_BADGE_SIZE_SM, FRAMEWORK_BADGE_SIZE_MD];

export const ROUTE_VIOLATIONS = 'violations';
export const ROUTE_FRAMEWORKS = 'frameworks';
export const TABS = [ROUTE_VIOLATIONS, ROUTE_FRAMEWORKS];

export const FRAMEWORKS_FILTER_TYPE_PROJECT = 'project';
export const FRAMEWORKS_FILTER_TYPE_FRAMEWORK = 'framework';
export const FRAMEWORKS_FILTER_VALUE_NO_FRAMEWORK = {
  color: '#000000',
  default: false,
  description: s__('ComplianceFrameworks|No framework'),
  id: 'gid://gitlab/ComplianceManagement::Framework/0',
  name: s__('ComplianceFrameworks|No framework'),
  pipelineConfigurationFullPath: '',
  __typename: 'ComplianceFramework',
};
