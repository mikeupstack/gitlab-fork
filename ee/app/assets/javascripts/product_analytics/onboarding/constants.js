import { s__, __ } from '~/locale';

export const INSTALL_NPM_PACKAGE = `yarn add @gitlab/application-sdk-js

--

npm i @gitlab/application-sdk-js`;

export const ESM_SETUP_WITH_NPM = `import { glClientSDK } from '@gitlab/application-sdk-js';

this.glClient = glClientSDK({ '$applicationId', '$host' });`;

export const COMMON_JS_SETUP_WITH_NPM = `const { glClientSDK } = require('@gitlab/application-sdk-js');

this.glClient = glClientSDK({ '$applicationId', '$host' });`;

export const HTML_SCRIPT_SETUP = `<script src="https://unpkg.com/@gitlab/application-sdk-js/dist/gl-sdk.min.js"></script>
<script>window.glClient = window.glSDK.glClientSDK({
    applicationId: '$applicationId',
    host: '$host',
});</script>`;

export const SHORT_POLLING_INTERVAL = 1000;

export const LONG_POLLING_INTERVAL = 2500;

export const STATE_CREATE_INSTANCE = 'CREATE_INSTANCE';

export const STATE_LOADING_INSTANCE = 'LOADING_INSTANCE';

export const STATE_WAITING_FOR_EVENTS = 'WAITING_FOR_EVENTS';

export const STATE_COMPLETE = 'COMPLETE';

export const EMPTY_STATE_I18N = {
  empty: {
    title: s__('ProductAnalytics|Analyze your product with Product Analytics'),
    description: s__(
      'ProductAnalytics|Set up Product Analytics to track how your product is performing. Combine it with your GitLab data to better understand where you can improve your product and development processes.',
    ),
    setUpBtnText: s__('ProductAnalytics|Set up product analytics'),
    learnMoreBtnText: __('Learn more'),
  },
  loading: {
    title: s__('ProductAnalytics|Creating your product analytics instance...'),
    description: s__(
      'ProductAnalytics|This might take a while, feel free to navigate away from this page and come back later.',
    ),
  },
};

export const FETCH_ERROR_MESSAGE = s__(
  'ProductAnalytics|An error occurred while fetching data. Refresh the page to try again.',
);

export const ONBOARDING_LIST_ITEM_I18N = {
  title: __('Product Analytics'),
  description: s__(
    'ProductAnalytics|Set up to track how your product is performing and optimize your product and development processes.',
  ),
  waitingForEvents: s__('ProductAnalytics|Waiting for events'),
  loadingInstance: s__('ProductAnalytics|Loading instance'),
};
