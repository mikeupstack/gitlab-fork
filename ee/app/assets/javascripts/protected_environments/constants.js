export const INHERITED_GROUPS = '1';
export const DEPLOYER_RULE_KEY = 'deploy_access_levels';
export const APPROVER_RULE_KEY = 'approval_rules';
export const RULE_KEYS = [DEPLOYER_RULE_KEY, APPROVER_RULE_KEY];

export const ACCESS_LEVELS = {
  DEPLOY: 'deploy_access_levels',
};

export const LEVEL_TYPES = {
  ROLE: 'role',
  USER: 'user',
  GROUP: 'group',
};
