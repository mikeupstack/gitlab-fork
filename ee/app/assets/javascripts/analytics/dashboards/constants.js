import { s__, __ } from '~/locale';
import { helpPagePath } from '~/helpers/help_page_helper';
import { KEY_METRICS, DORA_METRICS } from '~/analytics/shared/constants';

export const UNITS = {
  COUNT: 'COUNT',
  DAYS: 'DAYS',
  PER_DAY: 'PER_DAY',
  PERCENT: 'PERCENT',
};

export const TABLE_METRICS = {
  [DORA_METRICS.DEPLOYMENT_FREQUENCY]: {
    label: s__('DORA4Metrics|Deployment Frequency'),
    units: UNITS.PER_DAY,
  },
  [DORA_METRICS.LEAD_TIME_FOR_CHANGES]: {
    label: s__('DORA4Metrics|Lead Time for Changes'),
    invertTrendColor: true,
    units: UNITS.DAYS,
  },
  [DORA_METRICS.TIME_TO_RESTORE_SERVICE]: {
    label: s__('DORA4Metrics|Time to Restore Service'),
    invertTrendColor: true,
    units: UNITS.DAYS,
  },
  [DORA_METRICS.CHANGE_FAILURE_RATE]: {
    label: s__('DORA4Metrics|Change Failure Rate'),
    invertTrendColor: true,
    units: UNITS.PERCENT,
  },
  [KEY_METRICS.LEAD_TIME]: {
    label: s__('DORA4Metrics|Lead time'),
    invertTrendColor: true,
    units: UNITS.DAYS,
  },
  [KEY_METRICS.CYCLE_TIME]: {
    label: s__('DORA4Metrics|Cycle time'),
    invertTrendColor: true,
    units: UNITS.DAYS,
  },
  [KEY_METRICS.ISSUES]: {
    label: s__('DORA4Metrics|New issues'),
    units: UNITS.COUNT,
  },
  [KEY_METRICS.DEPLOYS]: {
    label: s__('DORA4Metrics|Deploys'),
    units: UNITS.COUNT,
  },
};

export const DASHBOARD_TITLE = s__('DORA4Metrics|Value Streams Dashboard');
export const DASHBOARD_DESCRIPTION = s__(
  'DORA4Metrics|The Value Streams Dashboard allows all stakeholders from executives to individual contributors to identify trends, patterns, and opportunities for software development improvements.',
);
export const DASHBOARD_DOCS_LINK = helpPagePath('user/analytics/value_streams_dashboard');
export const DASHBOARD_DESCRIPTION_GROUP = s__('DORA4Metrics|Metrics comparison for %{name} group');
export const DASHBOARD_DESCRIPTION_PROJECT = s__(
  'DORA4Metrics|Metrics comparison for %{name} project',
);
export const DASHBOARD_NO_DATA = __('No data available');
export const DASHBOARD_LOADING_FAILURE = __('Failed to load');

export const CHART_GRADIENT = ['#499767', '#5252B5'];
export const CHART_GRADIENT_INVERTED = [...CHART_GRADIENT].reverse();
export const CHART_LOADING_FAILURE = s__('DORA4Metrics|Failed to load charts');

export const CHART_TOOLTIP_UNITS = {
  [UNITS.COUNT]: undefined,
  [UNITS.DAYS]: __('days'),
  [UNITS.PER_DAY]: __('/day'),
  [UNITS.PERCENT]: '%',
};
