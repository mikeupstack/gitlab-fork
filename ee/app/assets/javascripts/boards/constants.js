/* eslint-disable import/export */
import { FilterFields as FilterFieldsCE } from '~/boards/constants';
import destroyBoardListMutation from '~/boards/graphql/board_list_destroy.mutation.graphql';
import updateBoardListMutation from '~/boards/graphql/board_list_update.mutation.graphql';
import listIssuesQuery from '~/boards/graphql/lists_issues.query.graphql';
import { TYPE_EPIC, TYPE_ISSUE } from '~/issues/constants';
import { s__ } from '~/locale';

import boardListsQuery from './graphql/board_lists.query.graphql';
import destroyEpicBoardListMutation from './graphql/epic_board_list_destroy.mutation.graphql';
import updateEpicBoardListMutation from './graphql/epic_board_list_update.mutation.graphql';
import epicBoardListsQuery from './graphql/epic_board_lists.query.graphql';
import listEpicsQuery from './graphql/lists_epics.query.graphql';
import listEpicsWithColorQuery from './graphql/lists_epics_with_color.query.graphql';
import listDeferredQuery from './graphql/board_lists_deferred.query.graphql';
import epicListDeferredQuery from './graphql/epic_board_lists_deferred.query.graphql';

export * from '~/boards/constants';

export const DRAGGABLE_TAG = 'div';

export const EPIC_LANE_BASE_HEIGHT = 40;

export const GroupByParamType = {
  epic: 'epic',
};

/* eslint-disable @gitlab/require-i18n-strings */
export const EpicFilterType = {
  any: 'Any',
  none: 'None',
};

export const HealthStatusFilterType = {
  any: 'Any',
  none: 'None',
};

export const IterationFilterType = {
  any: 'Any',
  none: 'None',
  current: 'Current',
};

export const MilestoneFilterType = {
  any: 'Any',
  none: 'None',
};

export const WeightFilterType = {
  none: 'None',
};
/* eslint-enable @gitlab/require-i18n-strings */

export const FilterFields = {
  [TYPE_ISSUE]: [
    ...FilterFieldsCE[TYPE_ISSUE],
    'epicId',
    'epicWildcardId',
    'weight',
    'iterationId',
    'iterationTitle',
    'iterationWildcardId',
    'iterationCadenceId',
    'healthStatusFilter',
  ],
  [TYPE_EPIC]: ['authorUsername', 'labelName', 'search', 'myReactionEmoji'],
};

export const IterationIDs = {
  NONE: 'gid://gitlab/Iteration/0',
  ANY: 'gid://gitlab/Iteration/-1',
  CURRENT: 'gid://gitlab/Iteration/-4',
};

export const ANY_ITERATION = {
  id: 'gid://gitlab/Iteration/-1',
  title: s__('BoardScope|Any iteration'),
  iterationCadenceId: null,
};

export const NO_ITERATION = {
  id: 'gid://gitlab/Iteration/0',
  title: s__('BoardScope|No iteration'),
  iterationCadenceId: null,
};

export const CURRENT_ITERATION = {
  id: 'gid://gitlab/Iteration/-4',
  title: s__('BoardScope|Current iteration'),
  iterationCadenceId: null,
};

export const IterationsPreset = [ANY_ITERATION, NO_ITERATION, CURRENT_ITERATION];

export const MilestoneIDs = {
  NONE: 0,
  ANY: -1,
};

export const DONT_FILTER_MILESTONE = {
  id: null,
  title: s__("BoardScope|Don't filter milestone"),
};
export const ANY_MILESTONE = {
  id: 'gid://gitlab/Milestone/-1',
  title: s__('BoardScope|Any Milestone'),
};
export const NO_MILESTONE = {
  id: 'gid://gitlab/Milestone/0',
  title: s__('BoardScope|No milestone'),
};
export const UPCOMING_MILESTONE = {
  id: 'gid://gitlab/Milestone/-2',
  title: s__('BoardScope|Upcoming'),
};
export const STARTED_MILESTONE = {
  id: 'gid://gitlab/Milestone/-3',
  title: s__('BoardScope|Started'),
};

export const MilestonesPreset = [
  DONT_FILTER_MILESTONE,
  ANY_MILESTONE,
  NO_MILESTONE,
  UPCOMING_MILESTONE,
  STARTED_MILESTONE,
];

export const ANY_ASSIGNEE = {
  id: 'gid://gitlab/User/-1',
  name: s__('BoardScope|Any assignee'),
};
export const AssigneesPreset = [ANY_ASSIGNEE];

export const WeightIDs = {
  NONE: -2,
  ANY: -1,
};

export const ErrorMessages = {
  fetchIssueError: s__(
    'Boards|An error occurred while fetching the board issues. Please reload the page.',
  ),
  fetchEpicsError: s__(
    'Boards|An error occurred while fetching the board epics. Please reload the page.',
  ),
};

export const listsQuery = {
  [TYPE_ISSUE]: {
    query: boardListsQuery,
  },
  [TYPE_EPIC]: {
    query: epicBoardListsQuery,
  },
};

export const listsDeferredQuery = {
  [TYPE_ISSUE]: {
    query: listDeferredQuery,
  },
  [TYPE_EPIC]: {
    query: epicListDeferredQuery,
  },
};

export const updateListQueries = {
  [TYPE_ISSUE]: {
    mutation: updateBoardListMutation,
  },
  [TYPE_EPIC]: {
    mutation: updateEpicBoardListMutation,
  },
};

export const deleteListQueries = {
  [TYPE_ISSUE]: {
    mutation: destroyBoardListMutation,
  },
  [TYPE_EPIC]: {
    mutation: destroyEpicBoardListMutation,
  },
};

export const listIssuablesQueries = {
  [TYPE_ISSUE]: {
    query: listIssuesQuery,
  },
  [TYPE_EPIC]: {
    query: gon?.features?.epicColorHighlight ? listEpicsWithColorQuery : listEpicsQuery,
  },
};

export default {
  DRAGGABLE_TAG,
  EpicFilterType,
};
