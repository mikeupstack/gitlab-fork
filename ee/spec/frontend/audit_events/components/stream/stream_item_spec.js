import Vue from 'vue';
import VueApollo from 'vue-apollo';
import { GlDisclosureDropdown, GlFormInputGroup } from '@gitlab/ui';
import { createAlert } from '~/alert';
import { mountExtended } from 'helpers/vue_test_utils_helper';
import createMockApollo from 'helpers/mock_apollo_helper';
import waitForPromises from 'helpers/wait_for_promises';
import deleteExternalDestination from 'ee/audit_events/graphql/mutations/delete_external_destination.mutation.graphql';
import { AUDIT_STREAMS_NETWORK_ERRORS, STREAM_ITEMS_I18N } from 'ee/audit_events/constants';
import StreamItem from 'ee/audit_events/components/stream/stream_item.vue';
import StreamDestinationEditor from 'ee/audit_events/components/stream/stream_destination_editor.vue';
import {
  destinationDeleteMutationPopulator,
  mockExternalDestinations,
  mockFiltersOptions,
} from '../../mock_data';

jest.mock('~/alert');
Vue.use(VueApollo);

describe('StreamItem', () => {
  let wrapper;

  const destinationWithoutFilters = mockExternalDestinations[0];
  const destinationWithFilters = mockExternalDestinations[1];
  const defaultDeleteSpy = jest.fn().mockResolvedValue(destinationDeleteMutationPopulator());
  const showModalSpy = jest.fn();

  const createComponent = (props = {}, deleteExternalDestinationSpy = defaultDeleteSpy) => {
    const mockApollo = createMockApollo([
      [deleteExternalDestination, deleteExternalDestinationSpy],
    ]);
    wrapper = mountExtended(StreamItem, {
      attachTo: document.body,
      apolloProvider: mockApollo,
      propsData: {
        item: destinationWithoutFilters,
        groupEventFilters: mockFiltersOptions,
        ...props,
      },
      stubs: {
        StreamDestinationEditor: true,
        GlModal: {
          template: '<div class="modal-stub"><slot></slot></div>',
          methods: {
            show: showModalSpy,
          },
        },
      },
    });
  };

  const findEditButton = () => wrapper.findByTestId('edit-btn');
  const findDeleteButton = () => wrapper.findByTestId('delete-btn');
  const findViewButton = () => wrapper.findByTestId('view-btn');
  const findActionsDropdown = () => wrapper.findComponent(GlDisclosureDropdown);
  const findEditor = () => wrapper.findComponent(StreamDestinationEditor);
  const findFilterBadge = () => wrapper.findByTestId('filter-badge');

  afterEach(() => {
    createAlert.mockClear();
  });

  describe('render', () => {
    beforeEach(() => {
      createComponent();
    });

    it('should not show the editor', () => {
      expect(findEditor().exists()).toBe(false);
    });
  });

  describe('deleting', () => {
    it('should emit deleted on success operation', async () => {
      createComponent();
      const deleteBtn = findDeleteButton();
      await deleteBtn.trigger('click');

      expect(findActionsDropdown().props('loading')).toBe(true);
      await waitForPromises();

      expect(findActionsDropdown().props('loading')).toBe(false);
      expect(wrapper.emitted('deleted')).toBeDefined();
      expect(createAlert).not.toHaveBeenCalled();
    });

    it('should not emit deleted when backend error occurs', async () => {
      const errorMsg = 'Random Error message';
      const deleteExternalDestinationErrorSpy = jest
        .fn()
        .mockResolvedValue(destinationDeleteMutationPopulator([errorMsg]));
      createComponent({}, deleteExternalDestinationErrorSpy);
      const deleteBtn = findDeleteButton();

      await deleteBtn.trigger('click');

      expect(findActionsDropdown().props('loading')).toBe(true);
      await waitForPromises();

      expect(findActionsDropdown().props('loading')).toBe(false);
      expect(wrapper.emitted('deleted')).toBeUndefined();
      expect(wrapper.emitted('error')).toBeDefined();
      expect(createAlert).toHaveBeenCalledWith({
        message: errorMsg,
      });
    });

    it('should not emit deleted when network error occurs', async () => {
      const error = new Error('Network error');
      createComponent({}, jest.fn().mockRejectedValue(error));

      const deleteBtn = findDeleteButton();

      await deleteBtn.trigger('click');

      expect(findActionsDropdown().props('loading')).toBe(true);
      await waitForPromises();

      expect(findActionsDropdown().props('loading')).toBe(false);

      expect(wrapper.emitted('deleted')).toBeUndefined();
      expect(wrapper.emitted('error')).toBeDefined();
      expect(createAlert).toHaveBeenCalledWith({
        message: AUDIT_STREAMS_NETWORK_ERRORS.DELETING_ERROR,
        captureError: true,
        error,
      });
    });
  });

  describe('editing', () => {
    beforeEach(async () => {
      createComponent();
      await findEditButton().trigger('click');
    });

    it('should pass the item to the editor', () => {
      expect(findEditor().exists()).toBe(true);
      expect(findEditor().props('item')).toStrictEqual(mockExternalDestinations[0]);
    });

    it('should emit the updated event when the editor fires its update event', async () => {
      findEditor().vm.$emit('updated');
      await waitForPromises();

      expect(wrapper.emitted('updated')).toBeDefined();
      expect(findEditor().exists()).toBe(false);
    });

    it('should emit the error event when the editor fires its error event', () => {
      findEditor().vm.$emit('error');

      expect(wrapper.emitted('error')).toBeDefined();
      expect(findEditor().exists()).toBe(true);
    });

    it('should close the editor when the editor fires its cancel event', async () => {
      findEditor().vm.$emit('cancel');
      await waitForPromises();

      expect(findEditor().exists()).toBe(false);
    });
  });

  it('should show modal when viewing token', async () => {
    createComponent();

    await findViewButton().trigger('click');

    expect(showModalSpy).toHaveBeenCalled();
    expect(
      wrapper.findComponent('.modal-stub').findComponent(GlFormInputGroup).props('value'),
    ).toBe(destinationWithoutFilters.verificationToken);
  });

  describe('when an item has event filters', () => {
    beforeEach(() => {
      createComponent({ item: destinationWithFilters });
    });

    it('should show filter badge', () => {
      expect(findFilterBadge().text()).toBe(STREAM_ITEMS_I18N.FILTER_BADGE_LABEL);
      expect(findFilterBadge().attributes('id')).toBe(destinationWithFilters.id);
    });

    it('renders a popover', () => {
      expect(wrapper.findByTestId('filter-popover').element).toMatchSnapshot();
    });
  });

  describe('when an item has no filter', () => {
    beforeEach(() => {
      createComponent();
    });

    it('should not show filter badge', () => {
      expect(findFilterBadge().exists()).toBe(false);
    });
  });
});
