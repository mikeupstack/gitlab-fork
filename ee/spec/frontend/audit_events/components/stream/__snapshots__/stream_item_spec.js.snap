// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StreamItem when an item has event filters renders a popover 1`] = `
<div
  class="gl-popover"
  data-testid="filter-popover"
>
  Destination has filters applied. 
  <a
    class="gl-link"
    href="/help/administration/audit_event_streaming#event-type-filters"
    rel="noopener"
    target="_blank"
  >
    
              What are filters?
            
  </a>
</div>
`;
